traffic_ops/traffic_ops_golang/acme/acme_account.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	if acmeAccount.Email != nil && acmeAccount.Provider != nil {
+		var prevEmail string
+		var prevProvider string
+		err := tx.QueryRow("SELECT email, provider from acme_account where email = $1 and provider = $2", acmeAccount.Email, acmeAccount.Provider).Scan(&prevEmail, &prevProvider)
+		if err != nil && err != sql.ErrNoRows {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New(fmt.Sprintf("checking if acme account with email %s and provider %s exists: %v", *acmeAccount.Email, *acmeAccount.Provider, err.Error())))
+			return
+		}
+
+		if prevEmail != "" && prevProvider != "" {
+			api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New("acme account already exists"), nil)
+			return
+		}
+	}
+
+	resultRows, err := inf.Tx.NamedQuery(createQuery, acmeAccount)
+	if err != nil {
+		userErr, sysErr, errCode := api.ParseDBError(err)
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer resultRows.Close()
+
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+	}
+	if rowsAffected == 0 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("acme account create: no account was inserted"))
+		return
+	} else if rowsAffected > 1 {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("too many rows returned from acme account insert"))
+		return
+	}
+
+	alerts := tc.CreateAlerts(tc.SuccessLevel, "Acme account created")
+	api.WriteAlertsObj(w, r, http.StatusCreated, alerts, acmeAccount)
+
+	changeLogMsg := fmt.Sprintf("ACME ACCOUNT: %s %s, ACTION: created", *acmeAccount.Email, *acmeAccount.Provider)
+	api.CreateChangeLogRawTx(api.ApiChange, changeLogMsg, inf.User, tx)
+}
+
+func Update(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	tx := inf.Tx.Tx
+
+	var acmeAccount AcmeAccount
+	if err := json.NewDecoder(r.Body).Decode(&acmeAccount); err != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, err, nil)
+		return
+	}
+
+	if validErr := acmeAccount.ValidateUpdate(tx); validErr != nil {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, validErr, nil)
+		return
+	}
+
+	var prevAccount AcmeAccount
+	err := tx.QueryRow("SELECT email, private_key, uri, provider from acme_account where email = $1 and provider = $2", acmeAccount.Email, acmeAccount.Provider).Scan(&prevAccount.Email, &prevAccount.PrivateKey, &prevAccount.Uri, &prevAccount.Provider)
+	if err == sql.ErrNoRows {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, errors.New(fmt.Sprintf("acme account not found")), nil)
+		return
+	}
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New(fmt.Sprintf("checking if acme account with email %s and provider %s exists: %v", *acmeAccount.Email, *acmeAccount.Provider, err.Error())))
+		return
+	}
+
+	if acmeAccount.Uri == nil {
```

Comment:
`PUT` requests should be idempotent; by accepting `nil` here you're making this act like a `PATCH`. `PUT`s should require a *full* representation of the object being modified.